pandas/_libs/pd_parser.c

```diff
@@ -0,0 +1,178 @@
+        *result = HUGE_VAL;
+        *maybe_int = 0;
+      } else {
+        goto parsingerror;
+      }
+    } else if (strlen(data) == 8) {
+      if (0 == strcasecmp(data, "infinity")) {
+        *result = HUGE_VAL;
+        *maybe_int = 0;
+      } else {
+        goto parsingerror;
+      }
+    } else if (strlen(data) == 9) {
+      if (0 == strcasecmp(data, "-infinity")) {
+        *result = -HUGE_VAL;
+        *maybe_int = 0;
+      } else if (0 == strcasecmp(data, "+infinity")) {
+        *result = HUGE_VAL;
+        *maybe_int = 0;
+      } else {
+        goto parsingerror;
+      }
+    } else {
+      goto parsingerror;
+    }
+  }
+
+  Py_XDECREF(tmp);
+  return 0;
+
+parsingerror:
+  PyErr_Format(PyExc_ValueError, "Unable to parse string \"%s\"", data);
+  Py_XDECREF(tmp);
+  return -1;
+}
+
+
+static void pandas_parser_destructor(PyObject *op) {
+  void *ptr = PyCapsule_GetPointer(op, PandasParser_CAPSULE_NAME);
+  PyMem_Free(ptr);
+}
+
+static int pandas_parser_exec(PyObject *module) {
+  PandasParser_CAPI *capi = PyMem_Malloc(sizeof(PandasParser_CAPI));
+  if (capi == NULL) {
+    PyErr_NoMemory();
+    return -1;
+  }
+
+  capi->to_double = to_double;
+  capi->floatify = floatify;
+  capi->new_rd_source = new_rd_source;
+  capi->del_rd_source = del_rd_source;
+  capi->buffer_rd_bytes = buffer_rd_bytes;
+  capi->uint_state_init = uint_state_init;
+  capi->uint64_conflict = uint64_conflict;
+  capi->coliter_setup = coliter_setup;
+  capi->parser_new = parser_new;
+  capi->parser_init = parser_init;
+  capi->parser_free = parser_free;
+  capi->parser_del = parser_del;
+  capi->parser_add_skiprow = parser_add_skiprow;
+  capi->parser_set_skipfirstnrows = parser_set_skipfirstnrows;
+  capi->parser_set_default_options = parser_set_default_options;
+  capi->parser_consume_rows = parser_consume_rows;
+  capi->parser_trim_buffers = parser_trim_buffers;
+  capi->tokenize_all_rows = tokenize_all_rows;
+  capi->tokenize_nrows = tokenize_nrows;
+  capi->str_to_int64 = str_to_int64;
+  capi->str_to_uint64 = str_to_uint64;
+  capi->xstrtod = xstrtod;
+  capi->precise_xstrtod = precise_xstrtod;
+  capi->round_trip = round_trip;
+  capi->to_boolean = to_boolean;
+
+  PyObject *capsule =
+      PyCapsule_New(capi, PandasParser_CAPSULE_NAME, pandas_parser_destructor);
+  if (capsule == NULL) {
+    PyMem_Free(capi);
+    return -1;
+  }
+
+  // Monkeypatch the top level pandas module to have an attribute for the
+  // C-API. This is required because Python capsules do not support setting
+  // this attribute on anything but the top level package. Ideally not
+  // done when cpython gh-6898 gets implemented
+  PyObject *pandas = PyImport_ImportModule("pandas");
+  if (!pandas) {
+    PyErr_SetString(PyExc_ImportError,
+                    "pd_parser.c could not import module pandas");
+    Py_DECREF(capsule);
+    return -1;
+  }
+
+  if (PyModule_AddObject(pandas, "_pandas_parser_CAPI", capsule) < 0) {
```

Comment:
I'm interested why the parser needs to have this exposed to Python (I can't see to ctrl+F anything useful out of the diff).

Reply:
I think this is just a requirement of the CPython implementation:

https://docs.python.org/3/extending/extending.html#providing-a-c-api-for-an-extension-module

Ideally we wouldn't add this to the pandas namespace and attach it solely to the extension module itself. From what I could see in the CPython source though I don't think that would be possible without https://github.com/python/cpython/pull/6898